decodableco/mongo-flink
# DE-3303 & DE-3322: exactly-once and synchronous deletes in mongoflink

## Description

### Approach
Take advantage of per-record idempotence to achieve exactly-once behavior. Deletes are idempotent in mongo – no specific exceptions need to be caught/ignored to ensure this behavior.

To implement deletes, I created a custom Document called CdcDocument. CdcDocument is simply just a Document with a delete flag. The flag is set in RowDataDocumentSerializer based on a record's RowKind and is referenced in CommittableCdcTransaction (renamed from CommittableUpsertTransaction). I also needed to create a custom org.bson.codecs.CdcDocumentCodec to serialize bson to a CdcDocument.

After this PR is merged, mongoflink will have full CDC support. Also, added Decodable-README.md with deployment steps as well as necessary codeartifact distribution profiles to be able to deploy mongoflink to our codeartifact instance.

### Validation
package mongoflink locally with the following changes:

- insert a crash before commit with 33% prob (test 1)
- insert a crash after commit with 33% prob (test 2)
- run the sink, observe

run the validation as follows:

in upsert mode (using mysql-cdc source)

- [x] run crash tests individually
- [x] run crash tests together
expectation: no data loss, no duplicate records

## Security Considerations

Are there any security or data concerns to consider? If so, please describe them below.
This may include, but is not limited to:

 - Potential SQL injection
 - Handling customer credentials
 - Sensitive information in log messages
 - Dependencies with known vulnerabilities
 - Other issues listed in the OWASP [Top 10](https://owasp.org/Top10/)

## Type of change
- [ ] Task - An internal-only task that has no visible customer effect.
- [x] New feature - New functionality that is user-visible.
- [ ] Improvement - Existing functionality works better than it did before.
- [ ] Bug fix - Something didn’t work as expected, but now it does.
- [ ] Others (please describe)

## Documentation

- [x] This pull request introduce a new feature

If yes, how is the feature documented?
- [x] Docs - Please provide [links to docs](https://docs.decodable.co/docs/connector-reference-mongodb)
- [ ] Javadocs
- [ ] Not documented

## Checklists
- [x] Pull request has a descriptive title linked with JIRA ticket in the format `DE-XXXX: Summary` 
  where DE is the JIRA project short code, XXXX is the number, and Summary is the JIRA summary.
- [x] Pull request has reviewers assigned.
